### PR TITLE
Make resuming generators unsafe instead of the creation of immovable generators

### LINKED ISSUE #47787: Immovable generators and unsafe block
Immovable generators require unsafe block, thus allowing every unsafe function inside generator to work. There is no way to say I want immovable generator, but its body should not be marked unsafe.

```rust
unsafe { 
    static || {
        yield;
        let a =  Box::from_raw(0 as *mut usize);
    }
};
```

It can probably be fixed by adding new `ImmovableGenerator` trait and making creation of immovable generators safe.

cc @Zoxc

### PR DESCRIPTION
**Pull request by Zoxc**
cc @withoutboats

Fixes #47787

### CODE REVIEW

**Comment by cramertj** [comment]
Awesome! Thanks @Zoxc. Looks like a few of the tests still need `unsafe` blocks around their calls to `resume`.

**Comment by Zoxc** [comment]
Travis passes now.

File: src/test/run-pass/generator/iterator-count.rs [hunk 1]
```diff
@@ -18,7 +18,7 @@ impl<T: Generator<Return = ()>> Iterator for W<T> {
     type Item = T::Yield;
 
     fn next(&mut self) -> Option<Self::Item> {
-        match self.0.resume() {
+        match unsafe { self.0.resume() } {
```
Comment (cramertj) [inline src/test/run-pass/generator/iterator-count.rs]: This impl isn't actually sound anymore since the iterator can move-- can you leave a comment?

**Comment by cramertj** [comment]
@bors r+

The `Iterator` for `W<impl Generator>` isn't sound anymore, but it's in a test (which correctly uses it without moving, so no UB will result) and will be replaced shortly, so I think it's fine to leave as-is.

**Comment by Zoxc** [comment]
@bors r=cramertj

@cramertj I already added a comment =P

**Comment by cramertj** [comment]
> I already added a comment =P

Nice :)

**Pull request merged**